debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/AttachedDataWrapperHelper.java

```diff
@@ -23,15 +23,18 @@
 
 import com.ibm.j9ddr.CorruptDataException;
 import com.ibm.j9ddr.vm29.pointer.U8Pointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9ShrOffSetPointer;
 import com.ibm.j9ddr.vm29.pointer.generated.AttachedDataWrapperPointer;
 import com.ibm.j9ddr.vm29.structure.AttachedDataWrapper;
 import com.ibm.j9ddr.vm29.structure.ShcItem;
 
 public class AttachedDataWrapperHelper {
-	// #define ADWCACHEOFFSET(adw) (((U_8*)(adw)) + J9SHR_READSRP((adw)->cacheOffset))
-	public static U8Pointer ADWCACHEOFFSET(AttachedDataWrapperPointer ptr) throws CorruptDataException {
-		return U8Pointer.cast(ptr).add(ptr.cacheOffset().longValue());
-	}	
+	public static U8Pointer ADWCACHEOFFSET(AttachedDataWrapperPointer ptr, U8Pointer[] cacheHeader) throws CorruptDataException {
+		if (J9ShrOffSetPointer.cast(ptr.cacheOffsetEA()).offset().eq(0)) {
+			return U8Pointer.NULL;
+		}
+		return cacheHeader[0].add(J9ShrOffSetPointer.cast(ptr.cacheOffsetEA()).offset());
```

Comment:
DDR needs to remain compatible with older caches. It will have to detect which type of cache is available and do the appropriate action.

Reply:
Fixed.